docs/proposals/super-simple-discovery-api.md

```diff
@@ -0,0 +1,205 @@
+
+#### Quick Primer on JWS and JWT
+
+[JSON Web Signatures](https://tools.ietf.org/html/rfc7515) are a way to sign, serialize and verify a payload.  It supports both symmetric keys (aka shared secrets) along with asymmetric keys (aka public key infrastructure or key pairs).  The JWS is split in to 3 parts:
+1. a header about how it is signed
+2. the clear text payload
+3. the signature.
+
+There are a couple of different ways of encoding this data -- either as a JSON object or as a set of BASE64URL strings for including in headers or URL parameters.  In this case, we are using a shared secret and the HMAC-SHA256 signing algorithm and encoding it as a JSON object.  The popular JWT (JSON Web Tokens) specification is a type of JWS.
+
+The JWS specification [describes how to encode](https://tools.ietf.org/html/rfc7515#appendix-F) "detached content".  In this way the signature is calculated as normal but the content isn't included in the signature.
+
+#### NEW: JWT authentication
+
+To avoid passing an unencoded token we will instead create a short lived JWT based on the token.  The JWT will be constructed as so:
+
+* **Header**
+  * **typ**: `JWT`. REQUIRED.
+  * **alg**: `HS256` for HMAC SHA-256. REQUIRED.
+  * **kid**: The token-id. REQUIRED.
+* **Claims**
+  * **sub**: The token-id. REQUIRED.
+  * **aud**: The long term globally unique cluster identifier.  This is not implemented yet so is OPTIONAL.
+  * **exp**: An expiration time for this JWT.  This must be no more than 2 minutes in the future. REQUIRED.
+  * **iat**: The time that the JWT was issued.  OPTIONAL.
+
+**Question:** The 2 minute value is picked to be short enough to discourage replay attacks without introducing the complexity of tracking nonces.  If we mandated the use of nonces, we'd have to track which nonces were used across HA instances.  Is there a better value than 2 minutes?  If we see clock skew greater or connectivity problems greater than 2 minutes we'll probably have other problems.
+
+#### NEW: `kube-public` namespace
+
+Kubernetes ConfigMaps are per-namespace and are generally only visible to principals that have read access on that namespace.  To create a config map that *everyone* can see, we introduce a new `kube-public` namespace.  This namespace, by convention, is readable by all authenticated users.
+
+#### NEW: JWS content signing
+
+When the JWT flow is used, we go one step further and sign the payload of the response using a JWS with detached content.
+
+We compute the JWS as normal (using the token provided in the JWT and `HS256`) and then remove the middle section that contains the BASE64URL encoded content.  This shortened JWS (which only has the JWS header and signature) is then put in a new HTTP header.  This header is `X-Kubernetes-JWS`.
+
+We use a "detached content" JWS to ease implementation.  This can be implemented a simple HTTP request filter without having to modify the subtle content negotiation mechanisms in the API server.
```

Comment:
It's not the token management system, really.  It's the bootstrapping logic, which sits above the core.

Put another way, what if I want to sign other config in the system?  Why wouldn't I use this approach if I were anyone except a special case?  There's nothing magic about this particular config - and the less magical it is on the server, the better. 

Another way:

1. We have a controller that periodically refreshes service tokens if/when they expire.  Should we write code in kubernetes, or write a service token expiration / refresh controllers.
2. We have TLS certs stored in secrets.  Should we write code in kubernetes, or write a control loop (no matter where it runs) that periodically refreshes them?

I think the layer violation would be having signing and tokens be special.  The proposal here actually splits them so there is a set of innate bootstrap secrets that are refreshed periodically, and then a completely separate mechanism that also signs the cluster result using these bootstrap secrets.  Really no different than service account controllers operating on tokens or dockercfg secrets.